File: frontend/src/services/reports.ts
import { load, loadOne } from './storage'
import type {
  Client, ClientBreakdown, Invoice, InvoiceSummary, LawyerProfile,
  Project, ProjectBreakdown, ReportData, TimeEntry,
} from '@/types'

export interface ReportParams {
  date_from: string
  date_to: string
  client_id?: number | null
}

export const reportsService = {
  get: (params: ReportParams): Promise<ReportData> => {
    const allEntries = load<TimeEntry>('time_entries')
    const allInvoices = load<Invoice>('invoices')
    const projects = load<Project>('projects')
    const clients = load<Client>('clients')
    const profile = loadOne<LawyerProfile>('profile')

    let entries = allEntries.filter(
      e => e.date >= params.date_from && e.date <= params.date_to
    )
    let invoices = allInvoices.filter(
      inv => inv.issue_date >= params.date_from && inv.issue_date <= params.date_to
    )

    if (params.client_id) {
      const clientProjectIds = new Set(
        projects.filter(p => p.client_id === params.client_id).map(p => p.id)
      )
      entries = entries.filter(e => clientProjectIds.has(e.project_id))
      invoices = invoices.filter(inv => inv.client_id === params.client_id)
    }

    // Group entries by client → project
    const clientMap = new Map<number, Map<number, { hours: number; amount: number; count: number }>>()
    for (const e of entries) {
      const project = projects.find(p => p.id === e.project_id)
      if (!project) continue
      const rate = parseFloat(project.hourly_rate ?? profile?.default_hourly_rate ?? '0')
      const hours = parseFloat(e.duration_hours)
      const amount = hours * rate
      if (!clientMap.has(project.client_id)) clientMap.set(project.client_id, new Map())
      const projMap = clientMap.get(project.client_id)!
      const existing = projMap.get(project.id) ?? { hours: 0, amount: 0, count: 0 }
      projMap.set(project.id, { hours: existing.hours + hours, amount: existing.amount + amount, count: existing.count + 1 })
    }

    const breakdown: ClientBreakdown[] = []
    for (const [clientId, projMap] of clientMap.entries()) {
      const client = clients.find(c => c.id === clientId)
      const projectBreakdowns: ProjectBreakdown[] = []
      let clientHours = 0
      let clientAmount = 0
      for (const [projectId, stats] of projMap.entries()) {
        const project = projects.find(p => p.id === projectId)
        projectBreakdowns.push({
          project_id: projectId,
          project_name: project?.name ?? '—',
          entries_count: stats.count,
          hours: stats.hours,
          amount: stats.amount,
        })
        clientHours += stats.hours
        clientAmount += stats.amount
      }
      breakdown.push({
        client_id: clientId,
        client_name: client?.name ?? '—',
        hours: clientHours,
        amount: clientAmount,
        projects: projectBreakdowns,
      })
    }

    const total_hours = breakdown.reduce((s, c) => s + c.hours, 0)
    const total_amount = breakdown.reduce((s, c) => s + c.amount, 0)

    const invoice_summary: InvoiceSummary = {
      count_total: invoices.length,
      count_paid: invoices.filter(i => i.status === 'paid').length,
      count_unpaid: invoices.filter(i => i.status !== 'paid').length,
      count_overdue: invoices.filter(i => i.status === 'overdue').length,
      total_invoiced: invoices.reduce((s, i) => s + parseFloat(i.total_amount), 0),
      total_paid: invoices.filter(i => i.status === 'paid').reduce((s, i) => s + parseFloat(i.total_amount), 0),
      total_unpaid: invoices.filter(i => i.status !== 'paid').reduce((s, i) => s + parseFloat(i.total_amount), 0),
    }

    return Promise.resolve({
      date_from: params.date_from,
      date_to: params.date_to,
      client_id: params.client_id ?? null,
      total_hours,
      total_amount,
      breakdown,
      invoice_summary,
    })
  },

  downloadPdf: async (params: ReportParams): Promise<void> => {
    const data = await reportsService.get(params)

    const allEntries = load<TimeEntry>('time_entries')
    const projects = load<Project>('projects')
    const storedProfiles = load<LawyerProfile>('profiles')
    const activeId = localStorage.getItem('billing_active_profile_id')
    const profile = (activeId ? storedProfiles.find(p => p.id === Number(activeId)) : null)
      ?? storedProfiles[0] ?? null

    let entries = allEntries.filter(
      e => e.date >= params.date_from && e.date <= params.date_to
    )
    if (params.client_id) {
      const ids = new Set(projects.filter(p => p.client_id === params.client_id).map(p => p.id))
      entries = entries.filter(e => ids.has(e.project_id))
    }
    entries.sort((a, b) => a.date.localeCompare(b.date))

    const fmtAmt = (n: number, cur = '₽') =>
      n.toLocaleString('ru-RU', { minimumFractionDigits: 2 }) + '\u00a0' + cur

    const clientRowsHtml = data.breakdown.map(c => {
      const projRowsHtml = c.projects.map(proj => {
        const project = projects.find(p => p.id === proj.project_id)
        const rate = parseFloat(project?.hourly_rate ?? profile?.default_hourly_rate ?? '0')
        const cur = project?.currency ?? 'RUB'
        const curSym = cur === 'USD' ? '$' : cur === 'EUR' ? '€' : '₽'

        const projEntries = entries.filter(e => e.project_id === proj.project_id)
        const entryRowsHtml = projEntries.map(e => {
          const h = parseFloat(e.duration_hours)
          const amt = h * rate
          return `<tr class="entry-row">
            <td class="td-date">${e.date}</td>
            <td class="td-desc">${e.description ?? '—'}</td>
            <td class="td-num">${h.toFixed(2)}</td>
            <td class="td-num">${rate > 0 ? fmtAmt(amt, curSym) : '—'}</td>
          </tr>`
        }).join('')

        return `<tr class="proj-hdr">
          <td colspan="4">${proj.project_name}&nbsp;<span style="font-weight:400;color:#777">(${proj.entries_count}\u00a0зап.)</span></td>
        </tr>
        ${entryRowsHtml}
        <tr class="proj-total">
          <td colspan="2" class="td-right">Итого по проекту:</td>
          <td class="td-num">${proj.hours.toFixed(2)}\u00a0ч</td>
          <td class="td-num">${rate > 0 ? fmtAmt(proj.amount, curSym) : '—'}</td>
        </tr>`
      }).join('')

      return `<tr class="client-hdr">
        <td colspan="4">${c.client_name}</td>
      </tr>
      ${projRowsHtml}
      <tr class="client-total">
        <td colspan="2" class="td-right">Итого по клиенту:</td>
        <td class="td-num">${c.hours.toFixed(2)}\u00a0ч</td>
        <td class="td-num">${fmtAmt(c.amount)}</td>
      </tr>`
    }).join('')

    const sigName = profile?.full_name ?? ''

    const html = `<!DOCTYPE html>
<html lang="ru"><head>
  <meta charset="utf-8">
  <title>Отчёт ${params.date_from} — ${params.date_to}</title>
  <style>
    body{font-family:Arial,sans-serif;padding:40px;color:#222;font-size:13px}
    h1{font-size:20px;margin:0 0 4px}
    .sub{color:#666;margin-bottom:6px}
    .lawyer{color:#444;font-size:12px;margin-bottom:24px}
    table{width:100%;border-collapse:collapse;margin-bottom:24px}
    th{background:#f0f0f0;padding:8px;text-align:left;font-size:12px;border-bottom:2px solid #ccc}
    td{padding:5px 8px;border-bottom:1px solid #eee;font-size:12px}
    .td-num{text-align:right;white-space:nowrap}
    .td-right{text-align:right;color:#666}
    .td-date{color:#555;width:88px}
    .td-desc{color:#333}
    .client-hdr td{background:#dbeafe;font-weight:700;font-size:14px;padding:8px}
    .proj-hdr td{background:#f3f4f6;font-weight:600;font-size:12px;color:#374151;padding:6px 8px}
    .proj-total td{background:#f9fafb;font-size:12px}
    .client-total td{background:#dbeafe;font-weight:600;font-size:12px}
    .entry-row td{font-size:12px}
    .summary{display:flex;gap:16px;margin-bottom:24px}
    .box{border:1px solid #ddd;border-radius:6px;padding:12px;flex:1}
    .box-label{font-size:11px;color:#999;margin-bottom:4px}
    .box-value{font-size:18px;font-weight:700}
    .grand-total{text-align:right;font-size:16px;font-weight:700;margin-bottom:32px}
    .sig-block{margin-top:48px;display:flex;justify-content:flex-end}
    .sig-line{text-align:center}
    .sig-underline{border-top:1px solid #222;width:220px;margin:0 auto}
    .sig-name{font-size:12px;color:#555;margin-top:6px}
    @media print{body{padding:20px}}
  </style>
</head><body>
  <h1>Отчёт по работе</h1>
  <div class="sub">${params.date_from} — ${params.date_to}</div>
  ${sigName ? `<div class="lawyer">${sigName}</div>` : ''}
  <div class="summary">
    <div class="box"><div class="box-label">Всего часов</div><div class="box-value">${data.total_hours.toFixed(2)}</div></div>
    <div class="box"><div class="box-label">Сумма (руб.)</div><div class="box-value">${fmtAmt(data.total_amount)}</div></div>
    <div class="box"><div class="box-label">Счетов выставлено</div><div class="box-value">${data.invoice_summary.count_total}</div></div>
  </div>
  <table>
    <thead><tr>
      <th>Дата</th><th>Описание работы</th>
      <th style="text-align:right">Часы</th><th style="text-align:right">Сумма</th>
    </tr></thead>
    <tbody>${clientRowsHtml}</tbody>
  </table>
  <div class="grand-total">Итого: ${data.total_hours.toFixed(2)}&nbsp;ч &nbsp;|&nbsp; ${fmtAmt(data.total_amount)}</div>
  ${sigName ? `<div class="sig-block">
    <div class="sig-line">
      <div class="sig-underline"></div>
      <div class="sig-name">${sigName}</div>
    </div>
  </div>` : ''}
</body></html>`

    const win = window.open('', '_blank')
    if (win) {
      win.document.write(html)
      win.document.close()
      setTimeout(() => win.print(), 300)
    }
  },
}

File: backend/app/api/routes/reports.py
"""Reports endpoint — time & billing breakdown + invoice summary."""

from __future__ import annotations

from datetime import date
from decimal import Decimal

from fastapi import APIRouter, Depends, Query
from fastapi.responses import Response
from pydantic import BaseModel
from sqlalchemy.orm import Session, joinedload, selectinload

from app.db.database import get_db
from app.models.client import Client
from app.models.enums import InvoiceStatus, TimeEntryStatus
from app.models.invoice import Invoice
from app.models.invoice_item import InvoiceItem
from app.models.lawyer_profile import LawyerProfile
from app.models.project import Project
from app.models.time_entry import TimeEntry
from app.pdf.report_generator import (
    InvoiceSummaryData,
    ReportClientRow,
    ReportData,
    ReportProjectRow,
    render_report_pdf,
)

router = APIRouter()


# ── Response schemas ───────────────────────────────────────────────────────────

class ProjectBreakdown(BaseModel):
    project_id: int
    project_name: str
    entries_count: int
    hours: float
    amount: float


class ClientBreakdown(BaseModel):
    client_id: int
    client_name: str
    hours: float
    amount: float
    projects: list[ProjectBreakdown]


class InvoiceSummary(BaseModel):
    count_total: int
    count_paid: int
    count_unpaid: int
    count_overdue: int
    total_invoiced: float
    total_paid: float
    total_unpaid: float


class ReportResponse(BaseModel):
    date_from: date
    date_to: date
    client_id: int | None
    total_hours: float
    total_amount: float
    breakdown: list[ClientBreakdown]
    invoice_summary: InvoiceSummary


# ── Helpers ────────────────────────────────────────────────────────────────────

def _build_report(
    db: Session,
    date_from: date,
    date_to: date,
    client_id: int | None,
) -> ReportResponse:
    # Get default rate for projects that don't have their own
    profile = db.query(LawyerProfile).first()
    default_rate = profile.default_hourly_rate if profile else Decimal("0")

    # Time entries in period
    entries_q = (
        db.query(TimeEntry)
        .options(joinedload(TimeEntry.project).joinedload(Project.client))
        .filter(TimeEntry.date >= date_from, TimeEntry.date <= date_to)
    )
    if client_id is not None:
        entries_q = entries_q.join(Project).filter(Project.client_id == client_id)

    entries = entries_q.all()

    # Group by client → project
    client_map: dict[int, dict] = {}
    for e in entries:
        if not e.project:
            continue
        project = e.project
        client = project.client
        if client is None:
            continue

        rate = project.hourly_rate if project.hourly_rate is not None else default_rate
        amount = float(e.duration_hours) * float(rate)

        cid = client.id
        pid = project.id

        if cid not in client_map:
            client_map[cid] = {
                "client_id": cid,
                "client_name": client.name,
                "hours": 0.0,
                "amount": 0.0,
                "projects": {},
            }
        if pid not in client_map[cid]["projects"]:
            client_map[cid]["projects"][pid] = {
                "project_id": pid,
                "project_name": project.name,
                "entries_count": 0,
                "hours": 0.0,
                "amount": 0.0,
            }

        client_map[cid]["hours"] += float(e.duration_hours)
        client_map[cid]["amount"] += amount
        client_map[cid]["projects"][pid]["hours"] += float(e.duration_hours)
        client_map[cid]["projects"][pid]["amount"] += amount
        client_map[cid]["projects"][pid]["entries_count"] += 1

    breakdown = [
        ClientBreakdown(
            client_id=c["client_id"],
            client_name=c["client_name"],
            hours=round(c["hours"], 1),
            amount=round(c["amount"], 2),
            projects=[
                ProjectBreakdown(
                    project_id=p["project_id"],
                    project_name=p["project_name"],
                    entries_count=p["entries_count"],
                    hours=round(p["hours"], 1),
                    amount=round(p["amount"], 2),
                )
                for p in sorted(
                    c["projects"].values(), key=lambda x: x["hours"], reverse=True
                )
            ],
        )
        for c in sorted(
            client_map.values(), key=lambda x: x["hours"], reverse=True
        )
    ]

    total_hours = round(sum(c.hours for c in breakdown), 1)
    total_amount = round(sum(c.amount for c in breakdown), 2)

    # Invoice summary in period
    inv_q = (
        db.query(Invoice)
        .options(selectinload(Invoice.items))
        .filter(Invoice.issue_date >= date_from, Invoice.issue_date <= date_to)
    )
    if client_id is not None:
        inv_q = inv_q.filter(Invoice.client_id == client_id)

    invoices = inv_q.all()
    today = date.today()

    count_paid = sum(1 for inv in invoices if inv.status == InvoiceStatus.paid)
    count_overdue = sum(
        1 for inv in invoices
        if inv.status == InvoiceStatus.overdue
        or (inv.status == InvoiceStatus.sent and inv.due_date < today)
    )
    count_unpaid = sum(
        1 for inv in invoices
        if inv.status in (InvoiceStatus.sent, InvoiceStatus.overdue, InvoiceStatus.draft)
    )
    count_total = len(invoices)

    total_invoiced = sum(
        sum(float(item.amount) for item in inv.items) for inv in invoices
    )
    total_paid = sum(
        sum(float(item.amount) for item in inv.items)
        for inv in invoices
        if inv.status == InvoiceStatus.paid
    )
    total_unpaid = sum(
        sum(float(item.amount) for item in inv.items)
        for inv in invoices
        if inv.status in (InvoiceStatus.sent, InvoiceStatus.overdue)
    )

    invoice_summary = InvoiceSummary(
        count_total=count_total,
        count_paid=count_paid,
        count_unpaid=count_unpaid,
        count_overdue=count_overdue,
        total_invoiced=round(total_invoiced, 2),
        total_paid=round(total_paid, 2),
        total_unpaid=round(total_unpaid, 2),
    )

    return ReportResponse(
        date_from=date_from,
        date_to=date_to,
        client_id=client_id,
        total_hours=total_hours,
        total_amount=total_amount,
        breakdown=breakdown,
        invoice_summary=invoice_summary,
    )


# ── Endpoints ──────────────────────────────────────────────────────────────────

@router.get("", response_model=ReportResponse, summary="Отчёт по времени и биллингу")
def get_report(
    date_from: date = Query(..., description="Начало периода"),
    date_to: date = Query(..., description="Конец периода"),
    client_id: int | None = Query(None, description="Фильтр по клиенту"),
    db: Session = Depends(get_db),
) -> ReportResponse:
    return _build_report(db, date_from, date_to, client_id)


@router.get(
    "/pdf",
    summary="Скачать отчёт в PDF",
    response_class=Response,
    responses={
        200: {"content": {"application/pdf": {}}, "description": "PDF-отчёт"},
    },
)
def get_report_pdf(
    date_from: date = Query(..., description="Начало периода"),
    date_to: date = Query(..., description="Конец периода"),
    client_id: int | None = Query(None, description="Фильтр по клиенту"),
    db: Session = Depends(get_db),
) -> Response:
    report_data = _build_report(db, date_from, date_to, client_id)

    # Resolve client name for PDF title
    client_name: str | None = None
    if client_id is not None:
        client = db.get(Client, client_id)
        if client:
            client_name = client.name

    pdf_report = ReportData(
        date_from=date_from,
        date_to=date_to,
        client_name=client_name,
        total_hours=report_data.total_hours,
        total_amount=report_data.total_amount,
        breakdown=[
            ReportClientRow(
                client_name=c.client_name,
                hours=c.hours,
                amount=c.amount,
                projects=[
                    ReportProjectRow(
                        project_name=p.project_name,
                        entries_count=p.entries_count,
                        hours=p.hours,
                        amount=p.amount,
                    )
                    for p in c.projects
                ],
            )
            for c in report_data.breakdown
        ],
        invoice_summary=InvoiceSummaryData(
            count_total=report_data.invoice_summary.count_total,
            count_paid=report_data.invoice_summary.count_paid,
            count_unpaid=report_data.invoice_summary.count_unpaid,
            count_overdue=report_data.invoice_summary.count_overdue,
            total_invoiced=report_data.invoice_summary.total_invoiced,
            total_paid=report_data.invoice_summary.total_paid,
            total_unpaid=report_data.invoice_summary.total_unpaid,
        ),
    )

    pdf_bytes = render_report_pdf(pdf_report)
    filename = f"report_{date_from}_{date_to}.pdf"
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )
